driver-core/src/main/com/mongodb/internal/Locks.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal;
+
+import com.mongodb.MongoInterruptedException;
+
+import java.util.concurrent.locks.Lock;
+import java.util.function.Supplier;
+
+/**
+ * This class is not part of the public API and may be removed or changed at any time.
+ */
+public final class Locks {
```

Comment:
Can be simplified down to:

```
public final class Locks {
    public static void withLock(final Lock lock, final Runnable action) {
       checkedWithLock(lock, () -> {
           action.run();
           return null;
       });
    }

    public static <V> V withLock(final Lock lock, final Supplier<V> supplier) {
        return checkedWithLock(lock, supplier::get);
    }

    public static <V, E extends Exception> V checkedWithLock(final Lock lock, final CheckedSupplier<V, E> supplier) throws E {
        try {
            lock.lockInterruptibly();
            try {
                return supplier.get();
            } finally {
                lock.unlock();
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new MongoInterruptedException("Interrupted waiting for lock", e);
        }
    }

    private Locks() {
    }
}
```

note: I simplified the method names as well.

Reply:
Much nicer.  I kind of liked the old names, but I see the appeal of these, so I changed them as suggested.

I'm not sure how the compiler is inferring that the `supplier::get` method reference is of type `CheckedSupplier<V, RuntimeException>` but that's what's happening so I won't argue.